File: packages/react-native-enriched-markdown/src/web/input/formatting/InputRemend.ts
interface DelimiterPair {
  open: string;
  close: string;
  symmetric: boolean;
}

// Longest first, so '***' matches before '**' and '**' before '*'.
const DELIMITER_PAIRS: DelimiterPair[] = [
  { open: '***', close: '***', symmetric: true },
  { open: '**', close: '**', symmetric: true },
  { open: '*', close: '*', symmetric: true },
  { open: '_', close: '_', symmetric: true },
  { open: '~~', close: '~~', symmetric: true },
  { open: '||', close: '||', symmetric: true },
  { open: '`', close: '`', symmetric: true },
  { open: '[', close: ']', symmetric: false },
];

function closingFor(entry: string): string {
  return DELIMITER_PAIRS.find((pair) => pair.open === entry)?.close ?? entry;
}

// Appends the closing delimiters missing from truncated or half-typed
// markdown, so the parser sees complete spans instead of literal delimiter
// characters.
export function completeMarkdown(markdown: string): string {
  if (markdown.length === 0) {
    return markdown;
  }

  const stack: string[] = [];
  let inLinkParen = false;
  const length = markdown.length;
  let i = 0;

  while (i < length) {
    const c = markdown.charAt(i);

    if (c === '\\' && i + 1 < length) {
      i += 2;
      continue;
    }

    // Link URL parentheses are a special two-character transition from "](".
    if (c === ']' && !inLinkParen && markdown.charAt(i + 1) === '(') {
      const bracketIndex = stack.lastIndexOf('[');
      if (bracketIndex !== -1) {
        // Styles left open inside the link text are contained by the bracket.
        stack.length = bracketIndex;
      }
      inLinkParen = true;
      i += 2;
      continue;
    }

    if (inLinkParen && c === ')') {
      inLinkParen = false;
      i++;
      continue;
    }

    if (inLinkParen) {
      i++;
      continue;
    }

    let matched = false;
    for (const pair of DELIMITER_PAIRS) {
      const openLen = pair.open.length;

      if (i + openLen > length) {
        continue;
      }

      if (pair.symmetric) {
        if (markdown.startsWith(pair.open, i)) {
          if (stack.length > 0 && stack[stack.length - 1] === pair.open) {
            stack.pop();
          } else {
            stack.push(pair.open);
          }
          i += openLen;
          matched = true;
          break;
        }
      } else {
        if (markdown.startsWith(pair.open, i)) {
          stack.push(pair.open);
          i += openLen;
          matched = true;
          break;
        }
        const closeLen = pair.close.length;
        if (i + closeLen <= length && markdown.startsWith(pair.close, i)) {
          if (stack.length > 0 && stack[stack.length - 1] === pair.open) {
            stack.pop();
          }
          i += closeLen;
          matched = true;
          break;
        }
      }
    }

    if (!matched) {
      i++;
    }
  }

  let suffix = '';

  if (inLinkParen) {
    suffix += ')';
  }

  for (let s = stack.length - 1; s >= 0; s--) {
    suffix += closingFor(stack[s]!);
  }

  if (suffix.length === 0) {
    return markdown;
  }

  return markdown + suffix;
}

File: android/src/main/java/com/swmansion/enriched/markdown/input/formatting/InputRemend.kt
package com.swmansion.enriched.markdown.input.formatting

object InputRemend {
  private data class DelimiterPair(
    val open: String,
    val close: String,
    val symmetric: Boolean,
  )

  private val DELIMITER_PAIRS =
    arrayOf(
      DelimiterPair("***", "***", true),
      DelimiterPair("**", "**", true),
      DelimiterPair("*", "*", true),
      DelimiterPair("_", "_", true),
      DelimiterPair("~~", "~~", true),
      DelimiterPair("||", "||", true),
      DelimiterPair("`", "`", true),
      DelimiterPair("[", "]", false),
    )

  fun complete(markdown: String): String {
    if (markdown.isEmpty()) return markdown

    val stack = mutableListOf<String>()
    var inLinkParen = false
    val length = markdown.length
    var i = 0

    while (i < length) {
      val c = markdown[i]

      if (c == '\\' && i + 1 < length) {
        i += 2
        continue
      }

      if (c == ']' && !inLinkParen && i + 1 < length && markdown[i + 1] == '(') {
        val bracketIndex = stack.lastIndexOf("[")
        if (bracketIndex != -1) {
          stack.subList(bracketIndex, stack.size).clear()
        }
        inLinkParen = true
        i += 2
        continue
      }

      if (inLinkParen && c == ')') {
        inLinkParen = false
        i++
        continue
      }

      if (inLinkParen) {
        i++
        continue
      }

      var matched = false
      for (pair in DELIMITER_PAIRS) {
        val openLen = pair.open.length

        if (i + openLen > length) continue

        val substring = markdown.substring(i, i + openLen)

        if (pair.symmetric) {
          if (substring == pair.open) {
            if (stack.isNotEmpty() && stack.last() == pair.open) {
              stack.removeAt(stack.lastIndex)
            } else {
              stack.add(pair.open)
            }
            i += openLen
            matched = true
            break
          }
        } else {
          if (substring == pair.open) {
            stack.add(pair.open)
            i += openLen
            matched = true
            break
          }
          val closeLen = pair.close.length
          if (i + closeLen <= length) {
            val closeSub = markdown.substring(i, i + closeLen)
            if (closeSub == pair.close) {
              if (stack.isNotEmpty() && stack.last() == pair.open) {
                stack.removeAt(stack.lastIndex)
              }
              i += closeLen
              matched = true
              break
            }
          }
        }
      }

      if (!matched) {
        i++
      }
    }

    val suffix = StringBuilder()

    if (inLinkParen) {
      suffix.append(")")
    }

    for (entry in stack.reversed()) {
      suffix.append(closingFor(entry))
    }

    if (suffix.isEmpty()) return markdown

    return markdown + suffix.toString()
  }

  private fun closingFor(entry: String): String = DELIMITER_PAIRS.firstOrNull { it.open == entry }?.close ?: entry
}
